File: apps/itsm/agentic/app/provision.py
"""weftd provisioning chain — Python port of the skill's ``provision.mjs``.

Lazily provisions the weftd app + toolset + skill + automation + graph on the
pre-existing tenant (``WEFT_TENANT_ID``) the first time ``ensure_app()`` runs.
The 5 weftd slugs + the model + the spec path + the display name are passed in
by the caller (``app/main.py`` resolves them — kept parameterized so this
module is the skill's generic Python template, not ITSM-hardcoded).

``weftd_api`` is sync (httpx.Client) — fine for the provisioning chain + the
session-creation call (both are short request/response, not the streaming
``/v1`` proxy — that's in ``session_routes.py`` with ``httpx.AsyncClient``).
"""
from __future__ import annotations

import json
import os
import threading
from pathlib import Path

import httpx

_HERE = Path(__file__).resolve().parent
_APP_ROOT = _HERE.parent  # apps/itsm/agentic/

_HTTP_METHODS = {"get", "post", "put", "patch", "delete", "head", "options"}


def derive_tool_names(spec: dict, toolset: str) -> list[str]:
    """Fail-closed allowlist: ``<toolset>_<operationId>``.lower() per operationId.

    the runtime's tool visibility is fail-closed on an empty allowlist — these
    ``<toolset>_*`` names must be listed for the named-toolset tools to be
    visible to the LLM. ``plan_route`` is runtime-internal, never listed.
    """
    names: list[str] = []
    for methods in (spec.get("paths") or {}).values():
        for m, op in (methods or {}).items():
            if m not in _HTTP_METHODS:
                continue
            if isinstance(op, dict) and op.get("operationId"):
                names.append(f"{toolset}_{op['operationId']}".lower())
    return names


class Provisioning:
    def __init__(self, *, weftd_base: str, api_key: str, tenant_id: str,
                 system_prompt: str, toolset: str, app_slug: str, app_name: str,
                 skill_slug: str, graph_slug: str, model: str, spec_path: str):
        self.weftd_base = weftd_base.rstrip("/")
        self.api_key = api_key
        self.tenant_id = tenant_id
        self.system_prompt = system_prompt
        self.toolset = toolset
        self.app_slug = app_slug
        self.app_name = app_name
        self.skill_slug = skill_slug
        self.graph_slug = graph_slug
        self.model = model
        spec = Path(spec_path)
        self.spec_path = spec if spec.is_absolute() else _APP_ROOT / spec
        self._ctx: dict | None = None
        self._lock = threading.Lock()
        self._client = httpx.Client(timeout=60.0, headers={
            "content-type": "application/json",
            "authorization": f"Bearer {api_key}",
        })

    def weftd_api(self, method: str, path: str, body=None) -> dict:
        """One provisioning call. Retries ONCE on a transient pre-handshake
        TLS/connect error (a reverse-proxy TLS-handshake drop — ``httpx.ConnectError`` /
        ``ConnectTimeout``), mirroring the ``/v1`` proxy's retry in
        ``session_routes.py``. HTTP 4xx/5xx are NOT retried — the provisioning
        chain is idempotent (find-or-create + draft/validate/publish/bind), so
        a re-run after a mid-chain failure is safe, but a single HTTP error is
        a real problem, not a transient one."""
        def _do() -> dict:
            resp = self._client.request(method, f"{self.weftd_base}{path}", json=body)
            if resp.status_code >= 400:
                try:
                    msg = resp.json().get("error", f"HTTP {resp.status_code}")
                except Exception:
                    msg = f"HTTP {resp.status_code}"
                raise RuntimeError(f"{msg} ({method} {path})")
            return resp.json() if resp.content else {}
        try:
            return _do()
        except (httpx.ConnectError, httpx.ConnectTimeout):
            return _do()  # one retry on a reverse-proxy TLS-handshake drop

    def _provision_app(self) -> dict:
        # tenant is pre-provisioned via the Weft console (WEFT_TENANT_ID).
        tenant_id = self.tenant_id
        if not tenant_id:
            t = self.weftd_api("POST", "/v1/tenants", {"name": self.app_name})
            tenant_id = t["tenant_id"]

        apps = self.weftd_api("GET", f"/v1/tenants/{tenant_id}/apps")
        existing = next(
            (a for a in (apps if isinstance(apps, list) else []) if a.get("slug") == self.app_slug),
            None,
        )
        # Register the control-plane app with the same origin allowlist the local
        # server enforces (ITSM_CORS_ORIGINS). Fall back to "*" only when unset —
        # the dev default — so a production integrator who sets ITSM_CORS_ORIGINS
        # gets weftd-side origin enforcement instead of an open wildcard.
        allowed_origins = [o.strip() for o in os.environ.get("ITSM_CORS_ORIGINS", "").split(",") if o.strip()]
        app = existing or self.weftd_api("POST", f"/v1/tenants/{tenant_id}/apps", {
            "slug": self.app_slug, "display_name": self.app_name,
            "allowed_origins": allowed_origins or ["*"],
        })
        app_id = app["app_id"]
        app_base = f"/v1/tenants/{tenant_id}/apps/{app_id}"

        spec = json.loads(self.spec_path.read_text())
        tool_names = derive_tool_names(spec, self.toolset)
        skill_payload = {
            "tool_names": tool_names,
            "system_prompt": self.system_prompt,
            "model": os.environ.get("OPENAI_MODEL") or self.model,
        }
        auto_payload = {"config_snapshot": json.dumps({
            "trigger": {"event": "session.created"},
            "action": {"type": "agent_message", "message":
                f"Hello! I'm the {self.app_name} assistant. I can help you manage "
                f"incidents, changes, and CIs. Try: 'P1: payment service down — "
                f"find the on-call, link the dependent CIs, assign + escalate'!"},
        })}

        # Toolset is tenant-scoped (sequential); skill/automation/graph are
        # app-scoped (independent — but kept sequential here for clarity).
        self._provision_toolset(tenant_id, spec)
        self._provision_skill(app_base, skill_payload)
        self._provision_automation(app_base, auto_payload)
        self._provision_graph(app_base)
        return {"tenant_id": tenant_id, "app_id": app_id}

    def _provision_toolset(self, tenant_id: str, spec: dict) -> None:
        # execution:"client" — same-origin + cookie + local → the browser
        # executes the tool calls (the cookie authenticates; X-Weft-Actor is
        # stamped by the runtime). base_url:"" → relative paths.
        self.weftd_api("PUT", f"/v1/tenants/{tenant_id}/toolsets/{self.toolset}", {
            "base_url": "", "execution": "client", "auth_type": "none",
            "spec": spec, "enabled": True,
        })

    def _provision_graph(self, app_base: str) -> None:
        graph_file = _APP_ROOT / f"{self.app_slug}-graph.json"
        if not graph_file.exists():
            print(f"[{self.app_slug}] {self.app_slug}-graph.json not found — "
                  f"skipping graph provisioning (generate + re-provision to bind)")
            return
        graph_text = graph_file.read_text()
        created = False
        try:
            self.weftd_api("POST", f"{app_base}/graphs", {
                "slug": self.graph_slug, "display_name": f"{self.app_name} Dependency Graph",
                "description": "Data-flow DAG for the API",
            })
            created = True
        except Exception:
            pass
        gv = self.weftd_api("PUT", f"{app_base}/graphs/{self.graph_slug}/draft", {"config_snapshot": graph_text})
        try:
            self.weftd_api("POST", f"{app_base}/graphs/{self.graph_slug}/validate")
            self.weftd_api("POST", f"{app_base}/graphs/{self.graph_slug}/publish")
        except Exception as e:
            # the analyzer's shared-field heuristic can produce data-flow
            # cycles (esp. when path params don't NAME-match the response id field —
            # e.g. {iid} vs `id` — so only spurious shared-field edges are inferred,
            # which cycle). Fail-open: skip the graph bind — the agent relies on the
            # system prompt + the reactive 409 backstop, NOT the graph.
            print(f"[{self.app_slug}] graph validate failed ({e}) — skipping bind "
                  f"(fail-open: the agent uses the system prompt + 409, not the graph)")
            return
        if not created:
            try:
                self.weftd_api("DELETE", f"{app_base}/graphs/{self.graph_slug}/unbind")
            except Exception:
                pass
        self.weftd_api("POST", f"{app_base}/graphs/{self.graph_slug}/bind", {"version_id": gv["version_id"]})

    def _provision_skill(self, app_base: str, payload: dict) -> None:
        created = False
        try:
            self.weftd_api("POST", f"{app_base}/skills", {
                "slug": self.skill_slug, "display_name": self.app_name, "description": "Assistant skill",
            })
            created = True
        except Exception:
            pass
        sv = self.weftd_api("PUT", f"{app_base}/skills/{self.skill_slug}/draft", payload)
        self.weftd_api("POST", f"{app_base}/skills/{self.skill_slug}/validate")
        self.weftd_api("POST", f"{app_base}/skills/{self.skill_slug}/publish")
        if not created:
            try:
                self.weftd_api("DELETE", f"{app_base}/skills/{self.skill_slug}/unbind")
            except Exception:
                pass
        self.weftd_api("POST", f"{app_base}/skills/{self.skill_slug}/bind", {"version_id": sv["version_id"]})

    def _provision_automation(self, app_base: str, payload: dict) -> None:
        created = False
        try:
            self.weftd_api("POST", f"{app_base}/automations", {
                "slug": "welcome", "display_name": "Welcome Greeting", "description": "Greeting",
            })
            created = True
        except Exception:
            pass
        av = self.weftd_api("PUT", f"{app_base}/automations/welcome/draft", payload)
        self.weftd_api("POST", f"{app_base}/automations/welcome/validate")
        self.weftd_api("POST", f"{app_base}/automations/welcome/publish")
        if not created:
            try:
                self.weftd_api("DELETE", f"{app_base}/automations/welcome/unbind")
            except Exception:
                pass
        self.weftd_api("POST", f"{app_base}/automations/welcome/bind", {"version_id": av["version_id"]})

    def ensure_app(self) -> dict:
        """Lazy + thread-safe. First call provisions; later calls return the ctx."""
        if self._ctx and self._ctx.get("app_id"):
            return self._ctx
        with self._lock:
            if self._ctx and self._ctx.get("app_id"):
                return self._ctx
            self._ctx = self._provision_app()  # raises on failure → _ctx stays None → next call retries
            return self._ctx


def create_provisioning(*, weftd_base, api_key, tenant_id, system_prompt,
                        toolset, app_slug, app_name, skill_slug, graph_slug, model, spec_path):
    return Provisioning(
        weftd_base=weftd_base, api_key=api_key, tenant_id=tenant_id,
        system_prompt=system_prompt, toolset=toolset, app_slug=app_slug,
        app_name=app_name, skill_slug=skill_slug, graph_slug=graph_slug,
        model=model, spec_path=spec_path,
    )

File: apps/online-store/agentic/server/provision.mjs
// weftd provisioning chain — parameterized extract of
// apps/store-agentic/server/shop-server.mjs (the weftdBase/apiKey/tenantId
// resolution + weftdAPI + provisionApp + provisionToolset + provisionGraph
// [ENOENT graceful-skip] + provisionSkill + provisionAutomation + ensureApp).
//
// The 5 weftd slugs + the model default + the spec path + the display name are
// placeholder tokens (rendered by scripts/scaffold.mjs in Task 3). The env
// resolution (WEFTD_BASE / WEFT_API_KEY / WEFT_TENANT_ID) lives in the CALLER —
// the scaffolded shop-server.mjs — which resolves them from process.env and
// passes the resolved values into createProvisioning(). This keeps the module
// zero-dep (no import of lib/system-prompt.mjs; the system prompt is injected
// as a string produced by scripts/gen-system-prompt.mjs in Task 7).
//
// Expected placement: server/provision.mjs (sibling of shop-server.mjs), so
// __dirname/.. resolves to the app root where openapi.json and
// online-store-agentic-graph.json live.

import { readFile } from 'node:fs/promises'
import { join } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = fileURLToPath(new URL('.', import.meta.url))

/**
 * Derive the fail-closed tool_names allowlist from the shop OpenAPI spec.
 *
 * Formula: for every operationId across all paths/methods in the spec,
 *   `shop_${operationId}`.toLowerCase().
 *
 * the runtime's tool visibility is fail-closed on an empty allowlist — the skill
 * must list these shop_* names for the named-toolset tools to be visible
 * to the LLM. plan_route is a runtime internal tool (always visible, no
 * listing).
 *
 * Phase 2b's scripts/gen-tool-allowlist.mjs (Task 4) supersedes this inline
 * derivation; it is preserved here to mirror v1's shape.
 *
 * @param {object} spec - Parsed OpenAPI spec (with `.paths`).
 * @returns {string[]} Ordered array of `shop_<operationId>`.toLowerCase() names.
 */
function deriveToolNames(spec) {
  const names = []
  for (const methods of Object.values(spec.paths ?? {})) {
    for (const [m, op] of Object.entries(methods ?? {})) {
      if (!['get', 'post', 'put', 'patch', 'delete', 'head', 'options'].includes(m)) continue
      if (op?.operationId) names.push(`shop_${op.operationId}`.toLowerCase())
    }
  }
  return names
}

/**
 * Lazily provisions the weftd app + toolset + skill + automation + graph on the
 * pre-existing tenant (WEFT_TENANT_ID) the first time ensureApp() is awaited.
 *
 * @param {object} opts
 * @param {string} opts.weftdBase     - Resolved weftd base URL (no trailing slash).
 * @param {string} opts.apiKey        - Tenant API key (resolved by the caller).
 * @param {string} opts.tenantId      - Tenant ID that owns the key (resolved by the caller).
 * @param {string} [opts.shopPort]    - Shop port (v1 used it to build shopBase for the
 *   system-prompt builder; under the named-toolset model that base arg is voided, so
 *   shopBase is inert — kept for shape parity with v1).
 * @param {string} opts.systemPrompt  - The agent system prompt string (produced by
 *   scripts/gen-system-prompt.mjs in Task 7). Injected to keep this module zero-dep.
 * @returns {{ ensureApp: () => Promise<{tenantId:string,appId:string}>, weftdAPI: (method:string,path:string,body?:object)=>Promise<object> }}
 */
export function createProvisioning({ weftdBase, apiKey, tenantId, shopPort, systemPrompt }) {
  // shopBase preserved from v1's shape (buildStoreSystemPrompt's base arg is
  // voided under the named-toolset model, so this is inert — kept for parity).
  const shopBase = shopPort ? `http://127.0.0.1:${shopPort}` : ''
  void shopBase

  // Lazily provisioned on first ensureApp(): { tenantId, appId }.
  let tenantCtx = null
  let provisionPromise = null

  function weftdAPI(method, path, body) {
    const headers = {
      'content-type': 'application/json',
      authorization: `Bearer ${apiKey}`,
    }
    // Retries ONCE on a transient pre-handshake TLS/connect error OR a mid-stream
    // stall (the reverse proxy in front of weftd intermittently drops the
    // TLS handshake / stalls mid-body — surfacing as `TypeError: fetch failed`
    // or an AbortError when the per-request timeout below fires). HTTP 4xx/5xx
    // are NOT retried — the provisioning chain is idempotent (find-or-create +
    // draft/validate/publish/bind), so a re-run after a mid-chain failure is
    // safe, but a single HTTP error is a real problem, not a transient one.
    // The 20s timeout is essential: WITHOUT it, a stalled fetch hangs
    // provisionApp forever → ensureApp() never settles → the chat bootstrap
    // (`POST /api/chat/session`) hangs indefinitely ("Connecting to chat…").
    const doFetch = (attempt = 0) =>
      fetch(`${weftdBase}${path}`, {
        method,
        headers,
        body: body !== undefined ? JSON.stringify(body) : undefined,
        signal: AbortSignal.timeout(20000),
      }).then(async (res) => {
        if (!res.ok) {
          let msg = `HTTP ${res.status}`
          let code
          try {
            // weftd nested: {"error":{"type","code","message"}} — prefer message,
            // keep code for logs. Flat legacy: {"error":"<code>","message":"…"}
            // (string error IS the code). Never assign an object to msg (→ "[object Object]").
            const e = await res.json()
            if (typeof e?.error === 'string') {
              code = e.error
              msg = e.message || e.error
            } else if (e?.error && typeof e.error === 'object') {
              code = typeof e.error.code === 'string' ? e.error.code : undefined
              msg = e.error.message || code || msg
            } else if (typeof e?.message === 'string') {
              msg = e.message
            }
          } catch {}
          const err = new Error(msg)
          err.status = res.status
          if (code) err.code = code
          throw err
        }
        return res.json()
      }).catch((err) => {
        // Retry on a transient connection error: a pre-handshake drop
        // (TypeError: fetch failed) OR a mid-stream stall that the 20s timeout
        // surfaced as AbortError/TimeoutError. The reverse proxy in front
        // of weftd intermittently stalls connections (notably for keep-alive
        // reuse from a long-running server process — a fresh short-lived
        // process is unaffected); a retry with a fresh connection almost
        // always succeeds. The provisioning chain is idempotent, so up to 3
        // retries (with backoff) is safe. HTTP 4xx/5xx already threw above
        // (not retried).
        const transient = err instanceof TypeError || err.name === 'TimeoutError' || err.name === 'AbortError'
        if (transient && attempt < 3) {
          return new Promise((resolve) => setTimeout(() => resolve(doFetch(attempt + 1)), 250 * (attempt + 1)))
        }
        throw err
      })
    return doFetch()
  }

  async function provisionApp() {
    // Remote mode: tenant is pre-provisioned via the Weft console (WEFT_TENANT_ID).
    if (tenantId) {
      tenantCtx = { tenantId, appId: null }
    } else {
      const t = await weftdAPI('POST', '/v1/tenants', { name: 'Online Store' })
      tenantCtx = { tenantId: t.tenant_id, appId: null }
    }

    // Find or create the online-store-agentic app (slug is unique per tenant).
    const existingApps = await weftdAPI('GET', `/v1/tenants/${tenantCtx.tenantId}/apps`)
    const existing = Array.isArray(existingApps) ? existingApps.find(a => a.slug === 'online-store-agentic') : null
    let app
    if (existing) {
      app = existing
    } else {
      // Register the control-plane app with the same origin allowlist the local
      // server enforces (DEMO_CORS_ORIGIN). Fall back to '*' only when unset —
      // the dev default — so a production integrator who sets DEMO_CORS_ORIGIN
      // gets weftd-side origin enforcement instead of an open wildcard.
      const allowedOrigins = (process.env.DEMO_CORS_ORIGIN ?? '')
        .split(',').map(o => o.trim()).filter(Boolean)
      app = await weftdAPI('POST', `/v1/tenants/${tenantCtx.tenantId}/apps`, {
        slug: 'online-store-agentic',
        display_name: 'Online Store',
        allowed_origins: allowedOrigins.length ? allowedOrigins : ['*'],
      })
    }
    tenantCtx.appId = app.app_id
    const appBase = `/v1/tenants/${tenantCtx.tenantId}/apps/${app.app_id}`

    // The named toolset provides the shop_* tools, but the runtime's
    // tool visibility is fail-closed on an empty allowlist — so the skill must
    // list the shop_* names (derived from the spec's operationIds) for
    // them to be visible to the LLM. plan_route is a runtime internal tool
    // (always visible, no listing needed).
    const specText = await readFile(join(__dirname, '..', 'openapi.json'), 'utf8')
    const toolsetSpec = JSON.parse(specText)
    // Fail-closed tool_names allowlist: every shop_<operationId>.toLowerCase()
    // derived from the spec. tools not listed here stay hidden.
    const toolNames = deriveToolNames(toolsetSpec)
    const skillPayload = {
      tool_names: toolNames,
      system_prompt: systemPrompt,
      model: process.env.OPENAI_MODEL || 'deepseek-v4-flash',
    }
    const autoPayload = {
      config_snapshot: JSON.stringify({
        trigger: { event: 'session.created' },
        action: { type: 'agent_message', message: "Hello! I'm the Online Store shopping assistant. I can help you browse products, manage your cart, and process orders. Try saying 'buy 2 mechanical keyboards and pay'!" },
      }),
    }

    // Skill/automation/graph provisioning are independent — run them in
    // parallel to cut the sequential call count. Each chain is: create
    // (idempotent) → draft → validate → publish → unbind (best-effort) → bind.
    // Toolset (tenant-level, named shop_* tools, execution:client) goes
    // first (sequential — it's tenant-scoped, not app-scoped).
    await provisionToolset(tenantCtx.tenantId, toolsetSpec)
    await Promise.all([
      provisionSkill(appBase, skillPayload),
      provisionAutomation(appBase, autoPayload),
      provisionGraph(appBase),
    ])

    return tenantCtx
  }

  // --- Reachability decision: execution:client for online-store-agentic ---
  //
  // The shop API is SAME-ORIGIN (the browser loads the page from 19745 and
  // calls /api/* + /v1 on 19745), COOKIE-AUTHED (the shop_session cookie
  // rides the same-origin fetch — no per-request credential), and LOCAL/PRIVATE
  // (binds 127.0.0.1, not reachable by the remote weftd). Together
  // that means the named shop_* tools must SUSPEND TO THE BROWSER: the
  // Weft runtime returns a tool-permission request to the chat panel, the
  // browser executes the call same-origin (the cookie authenticates it), and
  // the Weft runtime stamps X-Weft-Actor: agent on the outbound request. Hence
  // execution:"client".
  //
  // base_url:"" => relative paths: the tools follow whatever origin the page
  //   was loaded from. auth_type:"none" => the shop_session cookie handles auth
  //   on the same-origin fetch; no per-request bearer/sealed credential is
  //   attached to the toolset. X-Weft-Actor is the only header the runtime
  //   stamps.
  async function provisionToolset(tenantId, spec) {
    await weftdAPI('PUT', `/v1/tenants/${tenantId}/toolsets/shop`, {
      base_url: '',
      execution: 'client',
      auth_type: 'none',
      spec,
      enabled: true,
    })
  }

  // provisionGraph: a freshly-scaffolded app has NO online-store-agentic-graph.json yet
  // (Task 5's gen-graph.mjs generates it). If the file is absent, skip the
  // entire create/draft/validate/publish/bind chain — ENOENT graceful-skip so
  // first-provision doesn't reject (gap #2b). Task 5 generates the graph +
  // re-provisions to bind it.
  async function provisionGraph(appBase) {
    let graphText
    try {
      graphText = await readFile(join(__dirname, '..', 'online-store-agentic-graph.json'), 'utf8')
    } catch {
      console.warn('[online-store-agentic] online-store-agentic-graph.json not found — skipping graph provisioning (Task 5 generates + binds it)')
      return
    }
    let created = false
    try { await weftdAPI('POST', `${appBase}/graphs`, { slug: 'shopgraph', display_name: 'Online Store Dependency Graph', description: 'Data-flow DAG for the API' }); created = true } catch {}
    const graphVer = await weftdAPI('PUT', `${appBase}/graphs/shopgraph/draft`, { config_snapshot: graphText })
    await weftdAPI('POST', `${appBase}/graphs/shopgraph/validate`)
    await weftdAPI('POST', `${appBase}/graphs/shopgraph/publish`)
    if (!created) { try { await weftdAPI('DELETE', `${appBase}/graphs/shopgraph/unbind`) } catch {} }
    await weftdAPI('POST', `${appBase}/graphs/shopgraph/bind`, { version_id: graphVer.version_id })
  }

  async function provisionSkill(appBase, payload) {
    let created = false
    try { await weftdAPI('POST', `${appBase}/skills`, { slug: 'shopbot', display_name: 'Online Store', description: 'Assistant skill' }); created = true } catch {}
    const skillVer = await weftdAPI('PUT', `${appBase}/skills/shopbot/draft`, payload)
    await weftdAPI('POST', `${appBase}/skills/shopbot/validate`)
    await weftdAPI('POST', `${appBase}/skills/shopbot/publish`)
    // Skip unbind on first creation — nothing is bound yet.
    if (!created) { try { await weftdAPI('DELETE', `${appBase}/skills/shopbot/unbind`) } catch {} }
    await weftdAPI('POST', `${appBase}/skills/shopbot/bind`, { version_id: skillVer.version_id })
  }

  async function provisionAutomation(appBase, payload) {
    let created = false
    try { await weftdAPI('POST', `${appBase}/automations`, { slug: 'welcome', display_name: 'Welcome Greeting', description: 'Greeting' }); created = true } catch {}
    const autoVer = await weftdAPI('PUT', `${appBase}/automations/welcome/draft`, payload)
    await weftdAPI('POST', `${appBase}/automations/welcome/validate`)
    await weftdAPI('POST', `${appBase}/automations/welcome/publish`)
    if (!created) { try { await weftdAPI('DELETE', `${appBase}/automations/welcome/unbind`) } catch {} }
    await weftdAPI('POST', `${appBase}/automations/welcome/bind`, { version_id: autoVer.version_id })
  }

  async function ensureApp() {
    if (tenantCtx?.appId) return tenantCtx
    provisionPromise ??= provisionApp()
      .catch((err) => { provisionPromise = null; tenantCtx = null; throw err })
    return provisionPromise
  }

  return { ensureApp, weftdAPI }
}
